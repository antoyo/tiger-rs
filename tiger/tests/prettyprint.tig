let type tree = {key: string, left: tree, right: tree}
    function prettyprint(tree: tree): string =
        let var output := ""
            function write(s: string) =
                output := concat(output, s)
            function show(n: int, t: tree) =
                let function indent(s: string) =
                    (for i := 0 to n
                        do write(" ");
                        output := concat(output, s);
                        write("\n"))
                in
                    if t = nil then
                        indent(".")
                    else (
                        indent(t.key);
                        show(n + 4, t.left);
                        show(n + 4, t.right)
                    )
                end
        in
            show(0, tree);
            output
        end
in
    print(prettyprint(tree { key = "5", left = tree {
        key = "2", left = tree {
            key = "1", left = nil, right = nil
        }, right = tree {
            key = "3", left = nil, right = nil
        }
    }, right = tree {
        key = "7", left = tree {
            key = "6", left = nil, right = nil
        }, right = tree {
            key = "10", left = nil, right = nil
        }
    }}))
end
