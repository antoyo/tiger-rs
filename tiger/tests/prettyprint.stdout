 5
     2
         1
             .
             .
         3
             .
             .
     7
         6
             .
             .
         10
             .
             .
